### PR TITLE
Add enum classes

### LINKED ISSUE #43: [BUG] Enum classes not parsed
### Describe the bug
Enum classes do not appear to be parsable

### Requirements (place an `x` in each of the `[ ]`)**
* [x] I've read and understood the [Contributing guidelines](../CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

### To Reproduce
Using the example from https://docs.hhvm.com/hack/built-in-types/enum-class
```hack
enum class Random : mixed {
  int X = 42;
  string S = 'foo';
}
```
fails to parse.
Parse tree:
```query
enum_declaration [0, 0] - [3, 1]
  ERROR [0, 5] - [0, 10]
    identifier [0, 5] - [0, 10]
  name: identifier [0, 11] - [0, 17]
  type: type_specifier [0, 20] - [0, 25]
  enumerator [1, 2] - [1, 13]
    identifier [1, 2] - [1, 5]
    ERROR [1, 6] - [1, 7]
      identifier [1, 6] - [1, 7]
    integer [1, 10] - [1, 12]
  enumerator [2, 2] - [2, 19]
    identifier [2, 2] - [2, 8]
    ERROR [2, 9] - [2, 10]
      identifier [2, 9] - [2, 10]
    string [2, 13] - [2, 18]
```

### Expected behavior
Enum classes are correctly parsed

OS version(s): Linux

### PR DESCRIPTION
**Pull request by 4e554c4c**
Summary: This change adds support for enum classes (both abstract and regular)
to the parser. These differ quite a bit from ordinary enums and each-other; so
new rules have been written for each.

Fixes #43

Test Plan:
`tree-sitter test` succeeds

### Requirements (place an `x` in each `[ ]`)

* [x] I've added tests for any new code and ran `npm run test-corpus` to make sure all tests pass.
* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackhq/tree-sitter-hack/blob/main/.github/CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

**Comment by 4e554c4c** [comment]
Not sure why the validate test is failing. It seems to be running `bin/generate-corpus` and checking for new changes, but I cannot verify this on my end as there are no changes locally after running `bin/generate-corpus`.

File: grammar.js [hunk 1]
```diff
@@ -963,6 +965,44 @@ const rules = {
...
+      'abstract',
+      'enum',
+      'class',
+      field('name', $.identifier),
+      ':',
+      opt($.extends_clause),
+      field('type', $._type),
+      '{',
+      rep(choice(
+        seq($._type, $.identifier, '=', $._expression, ';'),
```
Comment (frankeld) [inline grammar.js:1001]: Where possible and makes sense, we like to use smaller rules to build larger rules. For example, we could use a variation of `seq($._type, $.enumerator)` here instead (may want to change/rename nodes as desired).

**Review by frankeld — COMMENTED** [review]
The validate task is failing because of the changes to expressions.txt. Generally, the sort order should be stable (`sort` should sort alphabetically by default) but your diff indicates otherwise. It may be an issue with whitespace/hyphens in the name, and could be probably be fixed by updating https://github.com/slackhq/tree-sitter-hack/blob/330470222a9abbaeaa09d2e7964aeb16b90b7a34/bin/generate-corpus#L18 to use the `-d` flag with `sort`. However, any insight into why your machine is using a different sort order than ubuntu and macos (where our tests run)?

**Comment by 4e554c4c** [comment]
> The validate task is failing because of the changes to expressions.txt. Generally, the sort order should be stable (`sort` should sort alphabetically by default) but your diff indicates otherwise. It may be an issue with whitespace/hyphens in the name, and could be probably be fixed by updating
> 
> https://github.com/slackhq/tree-sitter-hack/blob/330470222a9abbaeaa09d2e7964aeb16b90b7a34/bin/generate-corpus#L18
> to use the `-d` flag with `sort`. However, any insight into why your machine is using a different sort order than ubuntu and macos (where our tests run)?

ugh, I was thinking that this might be the case after that file changed. I'm using `sort (GNU coreutils) 8.32` on Fedora. Will try to make it match the test machines.... somehow.

**Comment by 4e554c4c** [comment]
well, i fixed the issues on my end by setting `LC_ALL="C"`. Not sure what's going on with the checks (is it also using my new `LC_ALL="C"` version of the script?)

**Comment by frankeld** [comment]
The error is coming from the change that adds `TSStateId *primary_state_ids` - you must be running a more recent version of tree-sitter. Our package.json entry implies that we are compatible with the 0.20.X versions but perhaps we should be using `~` instead of `^`. I can also try to upgrade to the more recent version.

**Review by frankeld — APPROVED** [review]
Will merge with https://github.com/slackhq/tree-sitter-hack/pull/46

**Pull request merged**